internal/app/machined/pkg/controllers/cluster/member.go

```diff
@@ -48,30 +50,28 @@ func (ctrl *MemberController) Outputs() []controller.Output {
 // Run implements controller.Controller interface.
 //
 //nolint:gocyclo
-func (ctrl *MemberController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+func (ctrl *MemberController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
 	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-r.EventCh():
+		if _, ok := channel.RecvWithContext(ctx, r.EventCh()); !ok && ctx.Err() != nil {
+			return nil //nolint:nilerr
 		}
 
-		affiliates, err := r.List(ctx, resource.NewMetadata(cluster.NamespaceName, cluster.AffiliateType, "", resource.VersionUndefined))
+		affiliates, err := safe.ReaderListAll[*cluster.Affiliate](ctx, r)
 		if err != nil {
 			return fmt.Errorf("error listing affiliates")
 		}
 
 		touchedIDs := make(map[resource.ID]struct{})
 
-		for _, affiliate := range affiliates.Items {
-			affiliateSpec := affiliate.(*cluster.Affiliate).TypedSpec()
+		for it := safe.IteratorFromList(affiliates); it.Next(); {
+			affiliateSpec := it.Value().TypedSpec()
 			if affiliateSpec.Nodename == "" {
 				// not a cluster member
 				continue
 			}
 
-			if err = r.Modify(ctx, cluster.NewMember(cluster.NamespaceName, affiliateSpec.Nodename), func(res resource.Resource) error {
-				spec := res.(*cluster.Member).TypedSpec()
+			if err = safe.WriterModify(ctx, r, cluster.NewMember(cluster.NamespaceName, affiliateSpec.Nodename), func(res *cluster.Member) error {
+				spec := res.TypedSpec()
 
 				spec.Addresses = slices.Clone(affiliateSpec.Addresses)
```

Comment:
I feel we should extract control plane's API server port from affiliates to members